tensorflow/contrib/go/graph.go

```diff
@@ -0,0 +1,570 @@
+// the provided graph string.
+func NewGraphFromText(graphStr string) (gr *Graph, err error) {
+	gr = NewGraph()
+	err = proto.UnmarshalText(graphStr, gr.def)
+
+	return
+}
+
+// LoadGraphFromFile Loads a Graph from the file on the specified path.
+func LoadGraphFromFile(path string) (gr *Graph, err error) {
+	graphStr, err := ioutil.ReadFile(path)
+	if err != nil {
+		return
+	}
+
+	gr = NewGraph()
+	err = proto.Unmarshal(graphStr, gr.def)
+
+	return
+}
+
+// LoadGraphFromTextFile Loads a Graph as plain text from the file on the specified
+// path.
+func LoadGraphFromTextFile(path string) (gr *Graph, err error) {
+	graphStr, err := ioutil.ReadFile(path)
+	if err != nil {
+		return
+	}
+
+	return NewGraphFromText(string(graphStr))
+}
+
+// Op Adds a new Node to the Graph with the specified operation, this function
+// could return an error if any of the mandatory attributes is not be present
+// or the value is not the expected for this attribute.
+func (gr *Graph) Op(opName string, name string, input []*GraphNode, device string, attrs map[string]interface{}) (node *GraphNode, err error) {
+	if err = gr.loadAvailableOps(); err != nil {
+		return
+	}
+
+	op, opFound := gr.availableOps[strings.ToLower(opName)]
+	if !opFound {
+		return nil, &ErrOperationNotFound{
+			op: opName,
+		}
+	}
+
+	if len(op.InputArg) != len(input) {
+		return nil, &ErrInvalidAmounthOfInputs{
+			operation:  opName,
+			opInputs:   len(op.InputArg),
+			specInputs: len(input),
+		}
+	}
+	inputs := make([]string, len(input))
+	for i, inNode := range input {
+		if op.InputArg[i].IsRef {
+			if inNode.ref == nil {
+				return nil, &ErrExpectedVarAsinput{
+					operation: opName,
+					inputPos:  i,
+				}
+			}
+			inputs[i] = inNode.ref.Name
+		} else {
+			inputs[i] = inNode.def.Name
+		}
+	}
+	node = &GraphNode{
+		def: &pb.NodeDef{
+			Name:   name,
+			Op:     opName,
+			Input:  inputs,
+			Device: device,
+			Attr:   make(map[string]*pb.AttrValue),
+		},
+		outDataTypes: make(map[string]DataType),
+	}
+
+	if attrs == nil {
+		attrs = make(map[string]interface{})
+	}
+	gr.matchTypes(input, node, attrs, op)
+
+	for _, attr := range op.Attr {
+		// Check if the attribute is specified, if it is not
+		// and doesn't have a default value, return an error since it
+		// is mandatory
+		if v, ok := attrs[attr.Name]; ok {
+			node.def.Attr[attr.Name] = gr.castAttrValue(attr.Type, v)
+			if node.def.Attr[attr.Name] == nil {
+				return nil, &ErrInvalidAttrValue{
+					operation:  opName,
+					attribName: attr.Name,
+				}
+			}
+		} else {
+			if attr.DefaultValue != nil {
+				node.def.Attr[attr.Name] = attr.DefaultValue
+			} else {
+				return nil, &ErrMandatoryAttributeNotSpecified{
+					operation:  opName,
+					attribName: attr.Name,
+				}
+			}
+		}
+	}
+
+	gr.def.Node = append(gr.def.Node, node.def)
+
+	return node, nil
+}
+
+// Variable Creates a variable operation and adds it to the graph. A variable
```

Comment:
Ups, yes, this is a bad habit that I have, I'm going to change the first capital letter in all the function comments.